plugins/Goals/Tracker/GoalsRequestProcessor.php

```diff
@@ -0,0 +1,136 @@
+use Piwik\Tracker\Request;
+use Piwik\Tracker\RequestProcessor;
+use Piwik\Tracker\Visit\VisitProperties;
+
+/**
+ * Handles conversion detection and tracking for tracker requests.
+ *
+ * ## Request Metadata
+ *
+ * This processor defines the following request metadata under the **Goals**
+ * plugin:
+ *
+ * * **goalsConverted**: The array of goals that were converted by this request. Each element
+ *                       will be an array of goal column value pairs. The ecommerce goal will
+ *                       only have the idgoal column set.
+ *
+ *                       Set in `processRequestParams()`.
+ *
+ *                       Plugins can set this to empty to skip conversion recording.
+ *
+ * * **visitIsConverted**: If `true`, the current visit should be marked as "converted". Note:
+ *                         some goal conversions (ie, ecommerce) do not mark the visit as
+ *                         "converted", so it is possible for goalsConverted to be non-empty
+ *                         while visitIsConverted is `false`.
+ *
+ *                         Set in `processRequestParams()`.
+ */
+class GoalsRequestProcessor extends RequestProcessor
+{
+    /**
+     * @var GoalManager
+     */
+    public $goalManager = null;
+
+    public function __construct(GoalManager $goalManager)
+    {
+        $this->goalManager = $goalManager;
+    }
+
+    public function processRequestParams(VisitProperties $visitProperties, Request $request)
+    {
+        $this->goalManager = new GoalManager();
+
+        if ($this->isManualGoalConversion($request)) {
+            // this request is from the JS call to piwikTracker.trackGoal()
+            $goal = $this->goalManager->detectGoalId($request->getIdSite(), $request);
+
+            $visitIsConverted = !empty($goal);
+            $visitProperties->setRequestMetadata('Goals', 'visitIsConverted', $visitIsConverted);
+
+            $existingConvertedGoals = $visitProperties->getRequestMetadata('Goals', 'goalsConverted') ?: array();
+            $visitProperties->setRequestMetadata('Goals', 'goalsConverted', array_merge($existingConvertedGoals, array($goal)));
+
+            $visitProperties->setRequestMetadata('Actions', 'action', null); // don't track actions when doing manual goal conversions
+
+            // if we find a idgoal in the URL, but then the goal is not valid, this is most likely a fake request
+            if (!$visitIsConverted) {
+                $idGoal = $request->getParam('idgoal');
+                Common::printDebug('Invalid goal tracking request for goal id = ' . $idGoal);
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public function afterRequestProcessed(VisitProperties $visitProperties, Request $request)
+    {
+        $goalsConverted = $visitProperties->getRequestMetadata('Goals', 'goalsConverted');
+
+        /** @var Action $action */
+        $action = $visitProperties->getRequestMetadata('Actions', 'action');
```

Comment:
Many plugin developers will have to do this I presume. Is there maybe a way to make this easier and more intuitive? At least just for actions?